contrib/platform/src/com/sun/jna/platform/win32/Advapi32Util.java

```diff
@@ -478,28 +478,45 @@ public static Account getTokenAccount(HANDLE hToken) {
 	 */
 	public static Account[] getCurrentUserGroups() {
 		HANDLEByReference phToken = new HANDLEByReference();
+		Win32Exception err = null;
 		try {
 			// open thread or process token
 			HANDLE threadHandle = Kernel32.INSTANCE.GetCurrentThread();
 			if (!Advapi32.INSTANCE.OpenThreadToken(threadHandle,
 					TOKEN_DUPLICATE | TOKEN_QUERY, true, phToken)) {
-				if (W32Errors.ERROR_NO_TOKEN != Kernel32.INSTANCE
-						.GetLastError()) {
-					throw new Win32Exception(Kernel32.INSTANCE.GetLastError());
+			    int rc = Kernel32.INSTANCE.GetLastError();
+				if (rc != W32Errors.ERROR_NO_TOKEN) {
+					throw new Win32Exception(rc);
 				}
+
 				HANDLE processHandle = Kernel32.INSTANCE.GetCurrentProcess();
 				if (!Advapi32.INSTANCE.OpenProcessToken(processHandle,
 						TOKEN_DUPLICATE | TOKEN_QUERY, phToken)) {
 					throw new Win32Exception(Kernel32.INSTANCE.GetLastError());
 				}
 			}
+
 			return getTokenGroups(phToken.getValue());
+		} catch(Win32Exception e) {
+	        err = e;
+		    throw err;    // re-throw in order to invoke finally block
 		} finally {
-			if (phToken.getValue() != WinBase.INVALID_HANDLE_VALUE) {
-				if (!Kernel32.INSTANCE.CloseHandle(phToken.getValue())) {
-					throw new Win32Exception(Kernel32.INSTANCE.GetLastError());
+		    HANDLE hToken = phToken.getValue();
+			if (!WinBase.INVALID_HANDLE_VALUE.equals(hToken)) {
+				try {
+				    Kernel32Util.closeHandle(hToken);
+				} catch(Win32Exception e) {
+				    if (err == null) {
+				        err = e;
+				    } else {
+				        err.addSuppressed(e);
+				    }
 				}
 			}
+
+			if (err != null) {
+			    throw err;
+			}
```

Comment:
I am sure nobody loves this whole way we handle closing and re-throwing, maybe there's a pattern we can implement part of a future PR where this gets wrapped into some boilerplate implementation ...